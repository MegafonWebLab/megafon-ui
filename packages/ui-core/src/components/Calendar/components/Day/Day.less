@import '~styles/base.less';

@b: e( '.mfui-beta-day');

@{b} {
    position: relative;
    .small();
    width: 32px;
    height: 32px;
    padding: 0 10px 8px 0;
    line-height: 24px;
    color: @freshAsphalt;
    border: none;
    background-color: rgba(0, 0, 0, 0);
    cursor: pointer;
    outline: none;
    text-decoration: none;
    box-sizing: content-box;
    -webkit-tap-highlight-color: rgba(0, 0, 0, 0);
    appearance: none;
    user-select: none;

    &:last-child {
        padding-right: 0px;
    }

    &_disabled {
        pointer-events: none;
        cursor: default;
        opacity: 0.3;
    }

    &_active,
    &:not(:disabled):not(&_active):focus
    {
        color: @clearWhite;
    }

    &:not(:disabled):not(&_active):hover:after,
    &:not(:disabled):not(&_active):focus:after,
    &_active:after
    {
        content: '';
        position: absolute;
        background-color: @spbSky0;
        color: @freshAsphalt;
        border-radius: 50% !important;
        width: 32px;
        height: 32px;
        top: 0;
        left: 0;
        z-index: 1;
    }


    &:not(:disabled):not(&_active):hover:after {
        background-color: @spbSky0;
        color: @freshAsphalt;
    }

    &:not(:disabled):not(&_active):focus:after,
    &_active:after
    {
        background-color: @green;
    }
    
    &_active + &_active:before,
    &_between:hover + &:not(&_monday)&_active:before
    {
        content: '';
        display: block;
        position: absolute;
        background-color: @spbSky1;
        height: 32px;
        width: 32px;
        top: 0;
        left: -22px;
        z-index: -1;
    }

    &_between {
        position: relative;
        border-radius: 0;

        &:before {
            content: '';
            display: block;
            position: absolute;
            z-index: -1;
            top: 0;
            left: -22px;
            padding: 0 20px;
            height: 32px;
            background-color: @spbSky1;
        }

        &:after {
            content: '';
            display: block;
            position: absolute;
            z-index: -1;
            top: 0;
            right: -16px;
            width: 40px;
            height: 32px;
            background-color: @spbSky1;
        }
    }

    &_sunday, &_last {
        border-radius: 50%;

        &:after {
            width: 32px;
            right: 10px;
            border-radius: 50%;
        }
    }

    &_last:after {
        right: 0;
    }

    &_monday, &_first {
        border-radius: 50%;

        &:before {
            padding: 0;
            border-radius: 50%;
            width: 32px;
            left: 0px;
        }
    }

    &_sunday&_first {
        border-radius: 50%;
    }

    &__label {
        position: relative;
        z-index: 99;
    }
}
