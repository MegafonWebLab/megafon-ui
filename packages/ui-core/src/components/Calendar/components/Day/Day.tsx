import React from 'react';
import PropTypes from 'prop-types';
import { useDay } from '@datepicker-react/hooks';
import cnCreate from 'utils/cnCreate';
import './Day.less';
import { isSunday, isMonday, isLastDayOfMonth, isFirstDayOfMonth } from 'date-fns';

export interface IDayPickerProps {
    isDateSelected: (date: Date) => boolean;
    isDateHovered: (date: Date) => boolean;
    isFirstOrLastSelectedDate: (date: Date) => boolean;
    isDateBlocked: (date: Date) => boolean;
    isDateFocused: (date: Date) => boolean;
    focusedDate: Date | null;
    onDateHover: (date: Date | null) => void;
    onDateSelect: (date: Date) => void;
    onDateFocus: (date: Date) => void;
}

export type DayType = {
    isBetween?: boolean;
    dayLabel?: string;
    date: Date;
    onMouseLeave?: () => void;
};

export type IDayProps = IDayPickerProps & DayType;

const cn = cnCreate('mfui-beta-day');
const Day: React.FC<IDayProps> = ({ isBetween = false, dayLabel, date, onMouseLeave, ...pickerProps }) => {
    const dayRef = React.useRef(null);

    const {
        isSelected,
        isSelectedStartOrEnd,
        disabledDate: isDateDisabled,
        onClick,
        onKeyDown,
        onMouseEnter,
        tabIndex,
    } = useDay({ date, dayRef, ...pickerProps });

    const modClasses = {
        active: isSelectedStartOrEnd,
        disabled: isDateDisabled,
        between: isBetween || (isSelected && !isSelectedStartOrEnd),
        sunday: isSunday(date),
        monday: isMonday(date),
        first: !isSelectedStartOrEnd && isFirstDayOfMonth(date),
        last: !isSelectedStartOrEnd && isLastDayOfMonth(date),
    };

    const handleMouseLeave = (): void => {
        onMouseLeave && onMouseLeave();
    };

    return (
        <button
            onClick={onClick}
            onKeyDown={onKeyDown}
            onMouseEnter={onMouseEnter}
            onMouseLeave={handleMouseLeave}
            tabIndex={tabIndex}
            type="button"
            ref={dayRef}
            className={cn(modClasses)}
        >
            <span className={cn('label')}>{dayLabel}</span>
        </button>
    );
};

Day.propTypes = {
    dayLabel: PropTypes.string,
    date: PropTypes.instanceOf(Date).isRequired,
    focusedDate: PropTypes.instanceOf(Date),
    isDateSelected: PropTypes.func.isRequired,
    isDateHovered: PropTypes.func.isRequired,
    isFirstOrLastSelectedDate: PropTypes.func.isRequired,
    isDateBlocked: PropTypes.func.isRequired,
    isDateFocused: PropTypes.func.isRequired,
    onDateHover: PropTypes.func.isRequired,
    onDateSelect: PropTypes.func.isRequired,
    onDateFocus: PropTypes.func.isRequired,
};

export default Day;
