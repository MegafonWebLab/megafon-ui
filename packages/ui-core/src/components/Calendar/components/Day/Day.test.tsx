import * as React from 'react';
import { shallow } from 'enzyme';
import Day, { IDayProps } from './Day';

const props: IDayProps = {
    dayLabel: 'Пн',
    date: new Date(1994, 9, 30),
    focusedDate: new Date(1994, 9, 30),
    isDateSelected: jest.fn(),
    isDateHovered: jest.fn(),
    isFirstOrLastSelectedDate: jest.fn(),
    isDateBlocked: jest.fn(),
    isDateFocused: jest.fn(),
    onDateHover: jest.fn(),
    onDateSelect: jest.fn(),
    onDateFocus: jest.fn(),
};

describe('<Day />', () => {
    describe('snapshots', () => {
        it('renders Day with required props', () => {
            const wrapper = shallow(<Day {...props} dayLabel={undefined}/>);

            expect(wrapper).toMatchSnapshot();
        });

        it('renders Day with all props', () => {
            const wrapper = shallow(<Day {...props} onMouseLeave={jest.fn()} isBetween />);

            expect(wrapper).toMatchSnapshot();
        });

        it('should return empty div if label is not provided', () => {
            const { dayLabel, ...restProps } = props;

            const wrapper = shallow(<Day {...restProps} />);

            expect(wrapper).toMatchSnapshot();
        });
    });

    describe('handleMouseLeave tests', () => {
        it('calls callback on them cursor leaves the button', () => {
            const onMouseLeave = jest.fn();

            const wrapper = shallow(<Day {...props} onMouseLeave={onMouseLeave} />);

            wrapper.find('button').simulate('mouseleave');

            expect(onMouseLeave).toBeCalled();
        });
    });
});
