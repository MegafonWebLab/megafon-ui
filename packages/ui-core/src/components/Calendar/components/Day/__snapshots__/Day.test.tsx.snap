// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Day /> snapshots renders Day with all props 1`] = `
<button
  className="mfui-beta-day mfui-beta-day_between mfui-beta-day_sunday"
  onClick={[Function]}
  onKeyDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  tabIndex={-1}
  type="button"
>
  <span
    className="mfui-beta-day__label"
  >
    Пн
  </span>
</button>
`;

exports[`<Day /> snapshots renders Day with required props 1`] = `
<button
  className="mfui-beta-day mfui-beta-day_sunday"
  onClick={[Function]}
  onKeyDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  tabIndex={-1}
  type="button"
>
  <span
    className="mfui-beta-day__label"
  />
</button>
`;

exports[`<Day /> snapshots should return empty div if label is not provided 1`] = `
<button
  className="mfui-beta-day mfui-beta-day_sunday"
  onClick={[Function]}
  onKeyDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  tabIndex={-1}
  type="button"
>
  <span
    className="mfui-beta-day__label"
  />
</button>
`;
